web3/contract.py

```diff
@@ -965,15 +916,53 @@ def factory(cls, **kwargs):
         if "method_name" not in kwargs:
             kwargs["method_name"] = cls.__name__
 
-        for key in kwargs:
-            if not hasattr(cls, key):
-                raise AttributeError(
-                    "Property {0} not found on ContractMethod class. "
-                    "`ContractMethod.factory` only accepts keyword arguments which are "
-                    "present on the ContractMethod class".format(key)
-                )
+        return PropertyCheckingFactory(kwargs["method_name"], (cls,), kwargs)
+
+
+class ContractEvent(object):
+    """Base class for contract events
+
+    An event accessed via the api contract.events.myEvents(*args, **kwargs)
+    is a subclass of this class.
+    """
+    address = None
+    event_name = None
+    web3 = None
+    contract_abi = None
+    abi = None
+
+    def __init__(self, *argument_names):
 
-        return type(kwargs["method_name"], (cls,), kwargs)
+        if argument_names is None:
+            self.argument_names = tuple()
+        else:
+            self.argument_names = argument_names
+
+        self.abi = self._get_event_abi()
+
+    @classmethod
+    def _get_event_abi(cls):
+        return find_matching_event_abi(
+            cls.contract_abi,
+            event_name=cls.event_name)
+
+    def processReceipt(self, txn_receipt):
+        return self._parse_logs(txn_receipt)
+
+    @to_tuple
+    def _parse_logs(self, txn_receipt):
+        for log in txn_receipt['logs']:
+            decoded_log = get_event_data(self.abi, log)
+            yield decoded_log
+
+    def _decode_log(self, log):
+        return get_event_data(self.abi, log)
+
+    @classmethod
+    def factory(cls, **kwargs):
+        if "event_name" not in kwargs:
+            kwargs["event_name"] = cls.__name__
```

Comment:
`kwargs.setdefault('event_name', cls.__name__)` is a slightly more concise way to do this.

Reply:
Alternatively, this could be made generic in `PropertyCheckingFactory` if we changed everything to use `class_name` for this argument.